src/metric_iterator.c

```diff
@@ -0,0 +1,119 @@
+#include "metric_iterator.h"
+#include "vector_index.h"
+
+static int MR_HasNext(void *ctx) {
+  MetricIterator *mr = ctx;
+  return mr->base.isValid;
+}
+
+static size_t MR_Len(void *ctx) {
+  MetricIterator *mr = ctx;
+  return mr->resultsNum;
+}
+
+static void MR_Abort(void *ctx) {
+  MetricIterator *mr = ctx;
+  mr->base.isValid = 0;
+}
+
+static t_docId MR_LastDocId(void *ctx) {
+  MetricIterator *mr = ctx;
+  return mr->lastDocId;
+}
+
+static void MR_Rewind(void *ctx) {
+  MetricIterator *mr = ctx;
+  mr->lastDocId = 0;
+  mr->curIndex = 0;
+  mr->base.isValid = 1;
+}
+
+static int MR_Read(void *ctx, RSIndexResult **hit) {
+  if (!MR_HasNext(ctx)) {
+    return INDEXREAD_EOF;
+  }
+  MetricIterator *mr = ctx;
+
+  // Set the item that we read in the current RSIndexResult
+  *hit = mr->base.current;
+  (*hit)->docId = mr->lastDocId = mr->idsList[mr->curIndex];
+  (*hit)->metric.value = mr->metricList[mr->curIndex];
+  (*hit)->metric.metricField = mr->fieldName;
+
+  mr->curIndex++;
+  if (mr->curIndex == mr->resultsNum) {
```

Comment:
add comments about advancing `curIndex` and reaching EOF. 

`IITER_SET_EOF` `IITER_CLEAR_EOF`